controllers/ingressnodefirewall_controller.go

```diff
@@ -217,28 +261,139 @@ func (r *IngressNodeFirewallReconciler) buildNodeStateSpecs(
+				continue withNextNode
 			}
-			if _, ok := nodeInterfaceSet[node.Name]; !ok {
-				nodeInterfaceSet[node.Name] = sets.String{}
+			// Otherwise, the current status will be either empty or ok, hence simply
+			// set it to SyncOK.
+			state.Status.SyncStatus = infv1alpha1.SyncOK
+
+			// Now, iterate over all interfaces in the InrgessNodeFirewallSpec.
+			if len(*firewallObj.Spec.Interfaces) == 0 {
+				state.Status = infv1alpha1.IngressNodeFirewallNodeStateStatus{
+					SyncStatus:       infv1alpha1.SyncError,
+					SyncErrorMessage: fmt.Sprintf("Invalid interface name - cannot provide an empty list"),
+				}
+				// Write back the state to the map and then continue with the next node.
+				nodeStates[node.Name] = state
+				continue withNextNode
 			}
-			nodeIngressRuleSet[node.Name] = append(nodeIngressRuleSet[node.Name], firewallObj.Spec.Ingress...)
-			if firewallObj.Spec.Interfaces != nil {
-				nodeInterfaceSet[node.Name].Insert(*firewallObj.Spec.Interfaces...)
+			for _, iface := range *firewallObj.Spec.Interfaces {
+				// Verify if an invalid interface name was specified.
+				// On error, report the error in the status field and continue with the next node.
+				if !isLegalInterfaceName(iface) {
+					state.Status = infv1alpha1.IngressNodeFirewallNodeStateStatus{
+						SyncStatus:       infv1alpha1.SyncError,
+						SyncErrorMessage: fmt.Sprintf("Invalid interface name %s", iface),
+					}
+					// Write back the state to the map and then continue with the next node.
+					nodeStates[node.Name] = state
+					continue withNextNode
+				}
+				// Create the rules for the node spec if they do not yet exist for this interface.
+				if _, ok := state.Spec.InterfaceIngressRules[iface]; !ok {
+					state.Spec.InterfaceIngressRules[iface] = []infv1alpha1.IngressNodeFirewallRules{}
+				}
+				// Merge in rules.
+				state.Spec.InterfaceIngressRules[iface], err = mergeRuleSet(
+					state.Spec.InterfaceIngressRules[iface], firewallObj.Spec.Ingress)
+				// On error, report the error in the status field and continue with the next node.
+				if err != nil {
+					state.Status = infv1alpha1.IngressNodeFirewallNodeStateStatus{
+						SyncStatus:       infv1alpha1.SyncError,
+						SyncErrorMessage: fmt.Sprintf("Illegal ruleset merge operation, err: %q", err),
+					}
+					// Write back the state to the map and then continue with the next node.
+					nodeStates[node.Name] = state
+					continue withNextNode
+				}
 			}
+			// Write back the state to the map.
+			nodeStates[node.Name] = state
 		}
 	}
 
-	nodeStateSpecs := make(map[string]infv1alpha1.IngressNodeFirewallNodeStateSpec, len(nodeIngressRuleSet))
-	for node, ingressRule := range nodeIngressRuleSet {
-		interfaceList := nodeInterfaceSet[node].List()
-		nodeStateSpecs[node] = infv1alpha1.IngressNodeFirewallNodeStateSpec{
-			Ingress:    ingressRule,
-			Interfaces: &interfaceList,
+	return nodeStates, nil
+}
+
+func isLegalInterfaceName(iface string) bool {
```

Comment:
we need also isValidInterface name which match one from `ip link show`

Reply:
Yeah, it has to go into the daemon though. In order to avoid further feature creep for this PR, I created: https://github.com/openshift/ingress-node-firewall/issues/91